tc and net probes refactoring

rafaeldtinoco opened the pull request:
## Initial Checklist

- [ ] There is an issue describing the need for this PR.
- [x] Git log contains summary of the change.
- [ ] Git log contains motivation and context of the change.
- [ ] If part of an EPIC, PR git log contains EPIC number.
- [ ] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

probes: refactor tc and network probes

- unify probes package for different probes types
- allow probes be attached/detached during runtime
- allow tc probes to be attached/detached from specific interface

This work will help in the future when dynamic attachment/detachment is
neded, but the current focus is to cleanup tracee.go initialization as
well (going forward with probes refactoring).

Commit adds:

- Probes interface:

  - Using a traceProbe Handle (types: rawTracepoint, kprobe, kretprobe)

      Syntax: Attach(EventHandle)
      Syntax: Detach(EventHandle)

  - Using a tcProbe Handle:

      Syntax: Attach(EventHandle, *net.Interface)
      Syntax: Detach(EventHandle, *net.Interface)

  - to detach all probes:

      DetachAll()

- Probe interface:

 - traceProbe (attach/detach/autoload)
 - tcProbe (attach/detach/autoload)

## Type of change

- [ ] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [x] Code refactor (code improvement and/or code removal)
- [ ] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## How Has This Been Tested?

Reproduce the test by running:

```
$ sudo ./dist/tracee-ebpf -o format:json -o option:parse-arguments -o option:detect-syscall -trace comm=bash -trace follow -trace event=net_packet -trace net=docker0 -capture net=docker0
```

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [x] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

rafaeldtinoco commented:
@NDStrahilevitz I'll start using you more for reviewing code if you don't mind.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -82,51 +84,117 @@ func Init(module *bpf.Module) (Probes, error) {
 			LoadElfPhdrs:               {event: "load_elf_phdrs", probeType: kprobe, fn: "trace_load_elf_phdrs"},
 			Filldir64:                  {event: "filldir64", probeType: kprobe, fn: "trace_filldir64"},
 		},
+		netProbes: []*probe{
```
Why not write these as standard probes?

rafaeldtinoco replied:
Yep, I thought about it.. but it felt weird to me we would have to add all the handles to each existing network event, something like:

<img width="557" alt="image" src="https://user-images.githubusercontent.com/7395852/173385728-06c11397-2d57-4374-b615-9a210329b7eb.png">

of course we can optimize it, just explaining where I went that direction, let me see what I can do.

rafaeldtinoco replied:
Anyway, I think that is the price we have to pay. Events such as DNS event wont have ICMP probes, for example. So I think we are good.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -7,26 +7,28 @@ import (
 type Probes interface {
+	// attach
 	Attach(probe Handle) error
+	AttachNetProbes() error
```
Can't we handle the net probes the same way we handle probes and just add them as event dependencies? Then we would load them using the same logic we do for standard probes. It would also help streamline the net events code in tracee.go

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -7,26 +7,28 @@ import (
 	Detach(probe Handle) error
+	DetachNetProbes() error
+	// no autoload
+	DisableNetProbesAutoload() error
```
It's reasonable to place this function in this package, but I don't think as part of the interface. I think this will occur naturally if we just treat the net probes the same way we treat the regular probes.
And if we do make the net probes part of the standard probes, this logic could possibly fit better in tracee itself (more of a stylistic decision but I do think there is some merit to keeping the interface minimal).

NDStrahilevitz commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,243 @@
+	AttachToInterfacesByName(ifaceNames []string) error
+	// detach
+	DetachFromInterface(iface *net.Interface) error
+	DetachFromInterfaceByName(ifaceName string) error
```
Same goes for these two as the above comment.

NDStrahilevitz reviewed (commented):
I think this refactor definitely does the job of cleaning up this code and making it more manageable, so I already like it just for that.

But I do think (at least for the net probes, for the tc probes it might be a bit more complex) it can be better integrated into the probe logics we have, which would make the net code more standardized to the other stuff we have and fit more naturally into the overall flow of code.

rafaeldtinoco commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,243 @@
+type TcProbes interface {
+	// attach
+	AttachToInterface(iface *net.Interface) error
+	AttachToInterfaceByName(ifaceName string) error
```
And there is also something to be added to what you said: we have to always index interfaces by ID only, and not name. I think I'll enforce the `iface *net.Interface` argument in the interface for now and solve altogether later (on were to do the interface name <-> *net.Interface conversion).

rafaeldtinoco commented:
I'm going to bump libbpfgo again because of:

https://github.com/aquasecurity/libbpfgo/pull/174

Which mitigates the following issue:

https://github.com/aquasecurity/tracee/issues/1676

So no exclusivity errors appear in tracee's output.

NDStrahilevitz commented on pkg/ebpf/events_definitions.go:
```diff
@@ -5428,6 +5442,15 @@ var EventsDefinitions = map[int32]EventDefinition{
 	DnsRequest: {
 		ID32Bit: sys32undefined,
 		Name:    "dns_request",
+		Probes: []probeDependency{
```
I actually think here we can keep the probes empty, same we do with container_create, since this is a derived event from NetPacket, it should declare the NetPacket event as a dependency.

rafaeldtinoco replied:
Example with dns_request only:

```
$ sudo ./dist/tracee-ebpf -o format:json -o option:parse-arguments -o option:detect-syscall -trace comm=bash -trace follow -trace event=dns_request -trace net=docker0 -capture net=docker0

{"timestamp":1655230060646770986,"processorId":0,"processId":13,"cgroupId":0,"threadId":13,"parentProcessId":9,"hostProcessId":3782544,"hostThreadId":3782544,"hostParentProcessId":3782540,"userId":0,"mountNamespace":4026533391,"pidNamespace":4026533430,"processName":"http","hostName":"","containerId":"c1ab59ed5848dee2c17cf334cf6c6a4bbfea73a9ead52bf58ef5f3f394a35501","containerImage":"","containerName":"","podName":"","podNamespace":"","podUID":"","eventId":"4001","eventName":"dns_request","argsNum":2,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"172.17.0.2","dst_ip":"1.1.1.1","src_port":33486,"dst_port":53,"protocol":17,"packet_len":78,"iface":"docker0"}},{"name":"dns_questions","type":"[]bufferdecoder.DnsQueryData","value":[{"query":"archive.ubuntu.com","queryType":"A","queryClass":"IN"}]}]}

{"timestamp":1655230060647083330,"processorId":0,"processId":12,"cgroupId":0,"threadId":12,"parentProcessId":9,"hostProcessId":3782543,"hostThreadId":3782543,"hostParentProcessId":3782540,"userId":0,"mountNamespace":4026533391,"pidNamespace":4026533430,"processName":"http","hostName":"","containerId":"c1ab59ed5848dee2c17cf334cf6c6a4bbfea73a9ead52bf58ef5f3f394a35501","containerImage":"","containerName":"","podName":"","podNamespace":"","podUID":"","eventId":"4001","eventName":"dns_request","argsNum":2,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"172.17.0.2","dst_ip":"1.1.1.1","src_port":34317,"dst_port":53,"protocol":17,"packet_len":79,"iface":"docker0"}},{"name":"dns_questions","type":"[]bufferdecoder.DnsQueryData","value":[{"query":"security.ubuntu.com","queryType":"A","queryClass":"IN"}]}]}
```

and with both:

```
$ sudo ./dist/tracee-ebpf -o format:json -o option:parse-arguments -o option:detect-syscall -trace comm=bash -trace follow -trace event=dns_request -trace event=net_packet -trace net=docker0 -capture net=docker0

{"timestamp":1655230109736500168,"processorId":0,"processId":215,"cgroupId":0,"threadId":215,"parentProcessId":211,"hostProcessId":3795107,"hostThreadId":3795107,"hostParentProcessId":3795103,"userId":0,"mountNamespace":4026533391,"pidNamespace":4026533430,"processName":"http","hostName":"","containerId":"c1ab59ed5848dee2c17cf334cf6c6a4bbfea73a9ead52bf58ef5f3f394a35501","containerImage":"","containerName":"","podName":"","podNamespace":"","podUID":"","eventId":"4000","eventName":"net_packet","argsNum":1,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"172.17.0.2","dst_ip":"1.1.1.1","src_port":33780,"dst_port":53,"protocol":17,"packet_len":78,"iface":"docker0"}}]}

{"timestamp":1655230109736500168,"processorId":0,"processId":215,"cgroupId":0,"threadId":215,"parentProcessId":211,"hostProcessId":3795107,"hostThreadId":3795107,"hostParentProcessId":3795103,"userId":0,"mountNamespace":4026533391,"pidNamespace":4026533430,"processName":"http","hostName":"","containerId":"c1ab59ed5848dee2c17cf334cf6c6a4bbfea73a9ead52bf58ef5f3f394a35501","containerImage":"","containerName":"","podName":"","podNamespace":"","podUID":"","eventId":"4001","eventName":"dns_request","argsNum":2,"returnValue":0,"stackAddresses":null,"args":[{"name":"metadata","type":"trace.PktMeta","value":{"src_ip":"172.17.0.2","dst_ip":"1.1.1.1","src_port":33780,"dst_port":53,"protocol":17,"packet_len":78,"iface":"docker0"}},{"name":"dns_questions","type":"[]bufferdecoder.DnsQueryData","value":[{"query":"archive.ubuntu.com","queryType":"A","queryClass":"IN"}]}]}
```

NDStrahilevitz replied:
You're right I misread the code. 
It does make me confused about the net_packet event now, since it seems the ebpf code can generate it, but it can also be derived inside the ebpf code into a dns_request for example. But then why would we derive it back? Obviously not a question for the scope of this PR though.

rafaeldtinoco replied:
Yep, it was done like that so the DNS event would also be capable of generating a net_event that generated it. Other way of approaching it would be to depend on net_event and to generate a DNS event from it (and others depending on the protocol handlers). Problem with that would be that every network event would depend on net_event (which has many probes). Let's talk about it, definitely.

NDStrahilevitz commented on pkg/ebpf/events_definitions.go:
```diff
@@ -5440,6 +5463,15 @@ var EventsDefinitions = map[int32]EventDefinition{
 	DnsResponse: {
 		ID32Bit: sys32undefined,
 		Name:    "dns_response",
+		Probes: []probeDependency{
```
Same as above here.

NDStrahilevitz commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,198 @@
+		return fmt.Errorf("could not find tc program %s", p.programName)
+	}
+
+	var tcOpts bpf.TcOpts
```
style nit: why not `tcOpts := bpf.TcOpts{ ProgFd: int(prog.GetFd()) }`?

rafaeldtinoco replied:
No special reason, I believe I just moved this code and did not touch it. I'll change it.

NDStrahilevitz reviewed (changes requested):
LGTM! 
But I did find one place where I'm almost sure there should be an error returned and not a print. Other than that I just have one style nit.

rafaeldtinoco commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	if module == nil {
+		fmt.Printf("eh null\n")
```
Just a leftover from dev. Removing it.

rafaeldtinoco commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,198 @@
+		err = p.tcHook.Destroy()
+	}
+
+	// TODO: check this
```
I still need to check this:

<img width="1214" alt="image" src="https://user-images.githubusercontent.com/7395852/173734197-34ae7bfd-f847-48f4-8eb9-da706af88832.png">

I don't want to leave a hanging program/filter just because we're afraid of deleting the qdisc. There is an upstream discussion about this that Yaniv started but, still, there are many BPF programs being left hanging there if you execute tracee multiple times.

rafaeldtinoco replied:
I'm opening the following issue for this: https://github.com/aquasecurity/tracee/issues/1828 AND will continue the exact same behavior in this refactoring.

rafaeldtinoco commented:
> LGTM! But I did find one place where I'm almost sure there should be an error returned and not a print. Other than that I just have one style nit.

Okay, I've done 2 fixes that were needed:

1. Tcprobe wasn't a pointer so detach had argument as nil.
2. Cleanup wasn't working correctly because of (1), made sure it is.
3. Opened a bug to address existing behavior issues

rafaeldtinoco commented:
@yanivagman I have requested @NDStrahilevitz to review this one for me (as he played a big part in the probe refactoring). I believe we are in good shape and I'm going to merge this soon, but wanted to highlight you first.

yanivagman commented:
@rafaeldtinoco if you don't mind, I would like to review this before you merge it.
Hope that I'll be able to do this today

rafaeldtinoco commented:
> @rafaeldtinoco if you don't mind, I would like to review this before you merge it. Hope that I'll be able to do this today

Nope, take your time. I highlighted you because of that. Thanks!

yanivagman reviewed (changes requested):
Indeed there's a lot of refactoring to be done in the network related code.
It was hard for me to review this PR as there is only one commit that does two different things:
1. Refactoring in the sense of moving code into new packages and between files
2. Changing logic - e.g. the way we destory hooks

I'm also in favor of making the tc probes "citizens with equal rights" and combine it with the regular probes by introducing a new attach type ("tc"). The specialized tc functions that deal with netwrok interfaces can be kept in a different file as it is done here though.

rafaeldtinoco commented on pkg/ebpf/events_definitions.go:
```diff
@@ -5428,6 +5442,15 @@ var EventsDefinitions = map[int32]EventDefinition{
+			{handle: probes.UDPDisconnect, required: true},
+			{handle: probes.UDPDestroySock, required: true},
+			{handle: probes.UDPv6DestroySock, required: true},
+			{handle: probes.InetSockSetState, required: true},
+			{handle: probes.Pingv4Sendmsg, required: true},
+			{handle: probes.Pingv6Sendmsg, required: true},
```
Nope, I should remove them, missed it.

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -7,26 +7,24 @@ import (
 	Attach(probe Handle) error
 	Detach(probe Handle) error
+	DetachAll() error
+	DisableNetAutoload() error
```
I thought about it as well. Let me see what I can do there.

rafaeldtinoco replied:
Okay, your idea was explained bellow saying that we could only autoload them all (but the tcProbes). I'll try to propose something else then.

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -7,26 +7,24 @@ import (
-	link      *bpf.BPFLink
+	module      *bpf.Module
+	probes      map[Handle]*probe
+	loadNetwork bool
```
<img width="374" alt="image" src="https://user-images.githubusercontent.com/7395852/174138932-e8a394a7-0990-43b1-b0d4-f74f0e706d77.png">
<img width="731" alt="image" src="https://user-images.githubusercontent.com/7395852/174138980-27aa8a03-e98f-40ca-8b00-053290768aaf.png">

rafaeldtinoco replied:
Im marking this as solved because the existence or not of this variable will depend on what we conclude in the previous comment (about autoloading network probes by default, during initialization, or not).

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -81,52 +79,108 @@ func Init(module *bpf.Module) (Probes, error) {
-//Detach the probe corresponding to the probe's handle
+// Detach detaches the probe corresponding to the probe's handle.
 func (p *probes) Detach(probe Handle) error {
+
```
I did some research about golang styling enforcement and could not find anything that called my attention (like clang-format). Let me know if you have any idea for style enforcement!

rafaeldtinoco replied:
I found this: https://github.com/uber-go/guide/blob/master/style.md but its huge and full of things we don't need. At the end they stick with golangci linter (which I already explored), I'll check that later. Let me know if there is a big blocker for you and I'll change styling.

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -81,52 +79,108 @@ func Init(module *bpf.Module) (Probes, error) {
+			ICMPv6Recv:                 {event: "icmpv6_rcv", probeType: kprobe, fn: "trace_icmpv6_rcv"},
+			ICMPv6Send:                 {event: "icmp6_send", probeType: kprobe, fn: "trace_icmp6_send"},
+			Pingv4Sendmsg:              {event: "ping_v4_sendmsg", probeType: kprobe, fn: "trace_ping_v4_sendmsg"},
+			Pingv6Sendmsg:              {event: "ping_v6_sendmsg", probeType: kprobe, fn: "trace_ping_v6_sendmsg"},
```
I agree, that is the part that Nadav also proposed (and I also felt when developing this). My comment to nadav was to merge this as is (better than what we currently have) and then bring tcProbes to Probes. If you think it is mandatory to have this now, we can try to have it now then (or to break into a next PR doing it). But I do agree.

rafaeldtinoco replied:
Ill propose something in this PR.

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -81,52 +79,108 @@ func Init(module *bpf.Module) (Probes, error) {
-		module: module,
-	}, nil
+		module:      module,
+		loadNetwork: true,
```
part of previous discussion.

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -81,52 +79,108 @@ func Init(module *bpf.Module) (Probes, error) {
-// Kprobes are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-kp
-// Tracepoints are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#3-tracep
-// Raw tracepoints are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#7-raw-tracep
+// DetachAll detaches the probe corresponding to the probe's handle.
```
Yep, thanks. It was left like that during development. Fixing.

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -81,52 +79,108 @@ func Init(module *bpf.Module) (Probes, error) {
+	return nil
+}
+
+// DisableNetAutoLoad disables eBPF prog auto load (and attach) for net probes
```
true.

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -81,52 +79,108 @@ func Init(module *bpf.Module) (Probes, error) {
+	p.loadNetwork = false
+
+	for handle, probe := range p.probes {
+		if int32(handle) >= 1000 { // network probes only
```
> We needed to disable network probes because some of the programs will require CAP_NET_ADMIN, and we don't want to force the user to have this capability if he doesn't need to use any network feature of tracee.
> Thinking about it, only the two tc programs we have require this capability, as the others are just kprobes.

But we are loading all eBPF programs by default now, no ? I know we can change that, I was just reflecting what we are doing in a refactor.

We're attaching them based on the dependencies, but we're loading them all. That is why I had the disable autoload for network probes (to accommodate what you said here). To avoid loading the network probes by default.

rafaeldtinoco replied:
> * I think that instead of looping all the probes, we can just disable the autoload for these two programs.

Ah, ok. I hope CAP_NET_ADMIN is not needed then.

> * If you choose to keep this code, I think it will be more clear to define "1000" as a constant. Also, it is prone to future bugs to have ">= 1000", as in the future someone may add handles starting from, e.g. 2000, and it will be disabled as well.

It sounds good to me if CAP_NET_ADMIN isn't needed. Actually it makes the network probes behave like regular probes, which I'm all in for.

rafaeldtinoco replied:
Yep, so you're right: Dropping net_admin capability I'm still able to load all probes (but not tcProbes):

If network filters are used:

<img width="1433" alt="image" src="https://user-images.githubusercontent.com/7395852/174156593-3bdbb389-27f7-417f-bd07-2ac54e54f7e2.png">

If not (but all regular network probes are still loaded):

<img width="1432" alt="image" src="https://user-images.githubusercontent.com/7395852/174156531-f6b48b9e-8756-414b-8e8c-57e8d632c416.png">

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -147,28 +201,48 @@ func (p *probe) attach(module *bpf.Module) error {
+// disableAutoload disables eBPF program automatic loading for given probe
+func (p *probe) disableAutoload(module *bpf.Module) error {
+
+	if p.link != nil { // already attached
+		return nil
+	}
```
true =D.

rafaeldtinoco commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,222 @@
+
+	// TODO: https://github.com/aquasecurity/tracee/issues/1828
+	// 1. entire clsact qdisc will be destroyed after this
+	// 2. Tc detach does not work and destroy needs to be used
```
It wasn't clear to me what it meant ("this check" wording, and 2 destroy
functions called made me confused). I had to re-read the mailing list
thread, and took sometime to find your email and answers related to it
(and see that your last email did not even had an answer upstream).

I prefer keeping like this if you don't mind, but I can add missing
information to that list if you think its appropriate.

IMO I would add anything else we judge important for the bug resolution,
including the huge mailing list URLs, in the issue I created (actually I would
even say its better to always create issues for things like that, and have a
quick explanation in the code pointing the URL).

rafaeldtinoco commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,222 @@
+	// 1. entire clsact qdisc will be destroyed after this
+	// 2. Tc detach does not work and destroy needs to be used
+
+	// tcOpts := bpf.TcOpts{}
+	// err = p.tcHook.Query(&tcOpts)
+	// if err != nil {
+	// 	return err
+	// }
+
+	// err = p.tcHook.Detach(&tcOpts)
+	// if err != nil {
+	// 	return err
+	// }
```
I wanted to show we should query tcopts, detach and destroy (if the API was fully working). I'll
move this to the case (and I do know that libbpf calls detach when destroy is called, either way
I tried detach before destroying to see if API was working.

rafaeldtinoco commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,222 @@
+	return err
+}
+
+// disableAutoload disables a tcProbe's program autoload when eBPF obj is loaded
+func (p *tcProbe) disableAutoload(module *bpf.Module) error {
+
+	prog, _ := module.GetProgram(p.programName)
+	if prog == nil {
+		return fmt.Errorf("could not find %s program", p.programName)
+	}
+
+	return prog.SetAutoload(false)
+}
```
It is, I'll move tcProbes into probes only like suggested.

rafaeldtinoco commented on pkg/ebpf/probes/tc_probes.go:
```diff
@@ -0,0 +1,222 @@
+			return err
+		}
+		if tcProbe.tcHook == nil {
+			fmt.Printf("HELL NO\n")
```
well, clearly a leftover from debugging =). fixing it.

rafaeldtinoco commented:
Moved to draft, still working in bringing tcProbes into Probes interface and declaring tcProbes as regular probe Dependencies (To all selected interfaces in that case).

rafaeldtinoco commented:
@NDStrahilevitz and I think probe should be an interface and both tracing probe and tc probe should implement it (it will be much cleaner than what it became by having both together). I had done that before this attempt but had doubts but after the result of not having an interface, and after a ping of him saying the same thing, I'll do that approach again).

rafaeldtinoco commented:
@NDStrahilevitz ok, so probes and tcprobes are part of the same package now. I have created 2 interfaces (Probes and Probe) and implemented Probe with "traceProbe" and "tcProbe".

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
-func (p *probes) Attach(probe Handle) error {
-	if p.probes[probe] == nil {
-		return fmt.Errorf("failed to attach probe: doesn't exist")
+func (p *probes) Attach(args ...interface{}) error {
```
![image](https://user-images.githubusercontent.com/7395852/175837961-1e83a829-8713-43d7-9212-fbf059e12605.png)

![image](https://user-images.githubusercontent.com/7395852/175839193-d9f94967-8e35-4f5f-a2d6-39bdc21b86a4.png)

![image](https://user-images.githubusercontent.com/7395852/175839204-487313a1-552e-45dc-af61-543560234aa7.png)

rafaeldtinoco commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
+	return p.attachOrDetach(false, args...)
+}
+
+func (p *probes) attachOrDetach(attach bool, args ...interface{}) error {
```
There was a discussion between Nadav and I about variadic functions being used here but I used it on purpose so the API could accept any argument (depending on existing and future probe types we may have).

I explained how to use the API here: https://github.com/aquasecurity/tracee/pull/1820/files#diff-a4767f3b581014c47b886c59c79fcf8374aee0edece50c13d6e9758df78abf53R12 and, despite preferences on implementation, the caller won't be affected by this decision.

For now user will attach a probe or detach it. User will also attach a tcProbe to an interface or detach it (and the API supports both, which is good enough and much better than what we currently have).

I'll the one using this interface to implement dynamic interface attachment/detachment and, if it does not serve its purpose, I could rethink it but, for now, I think it is good.

rafaeldtinoco commented:
I believe this is good for now, considering I'll be the one using this API for future work. Please let me know if there is a blocker or if we can go ahead and merge this for 0.7.1. Thanks!

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
-func (p *probe) attach(module *bpf.Module) error {
-	if p.link != nil {
-		return nil
+type Probe interface {
```
So this interface isn't the one you actually used for `traceProbe` and `tcProbe`, which DO have the same function signatures, just not the ones in the interface here.
I suggest we change the interface to fit what's actually used so we can use it in the `Probes` map, which could also remove a lot of the type casts (i've marked one such place with a review comment for example) making the code more readable.

So it should be (in pseudocode):
```
attach(args...)
detach(args...)
autoload(bool)
```

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
+	var err error
+
+	for _, in := range p.probes {
+		switch i := in.(type) {
```
For example: this type cast could be avoided if both implemented the interface

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
 type probes struct {
-	probes map[Handle]*probe
 	module *bpf.Module
+	probes map[EventHandle]interface{}
```
If we fix the interface this could be `map[EventHandle]probe`

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
+}
+
+// attach attaches an eBPF program to its probe
+func (p *traceProbe) attach(args ...interface{}) error {
```
so the in interface it should be:
`attach(args ...interface{})` and not `attach(module)`, same for detach and autoload.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
+}
+
+// attach attaches an eBPF program to its probe
+func (p *tcProbe) attach(args ...interface{}) error {
```
ditto

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
 	return err
 }
 
-type Handle int32
```
Not sure why the name change to EventHandle. 
Asking because the package explicitly removes itself from the events, so this was a deliberate naming choice (as in probeHandle -> probes.Handle).

yanivagman replied:
Yes, I wonder the same

rafaeldtinoco replied:
![image](https://user-images.githubusercontent.com/7395852/175932404-a092d040-2498-492d-98de-ad3c7a78844f.png)

NDStrahilevitz reviewed (commented):
So ideas for how to implement aside, I think there is a contradiction in the implementation here - that the singular probes don't implement the `probe` interface but a separate non existing code interface. So the written interface should be what's actually done, then it can be used in the package (i've listed examples).

To make it clear, i'm not suggesting to change how this is implemented now (contrary to what i've done in my PR to you), but to refine how it's done here.

Other than that, LGTM.

yanivagman commented on pkg/ebpf/tracee.go:
```diff
@@ -937,89 +838,41 @@ func (t *Tracee) initBPF() error {
 	}
 
+	// Open the eBPF object file (create a new module)
+
```
nit: why do we keep these extra lines between the comment and the code relevant to the comment? (here and in many other places)

rafaeldtinoco replied:
styling preferences. we should start enforcing one, clearly my preferences are different than yours, and different from Nadavs (who does comments like //something). =(

yanivagman commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,178 +2,503 @@ package probes
+		ICMPv6Recv:                 &traceProbe{eventName: "icmpv6_rcv", probeType: kprobe, programName: "trace_icmpv6_rcv", netProbe: true},
+		ICMPv6Send:                 &traceProbe{eventName: "icmp6_send", probeType: kprobe, programName: "trace_icmp6_send", netProbe: true},
+		Pingv4Sendmsg:              &traceProbe{eventName: "ping_v4_sendmsg", probeType: kprobe, programName: "trace_ping_v4_sendmsg"},
+		Pingv6Sendmsg:              &traceProbe{eventName: "ping_v6_sendmsg", probeType: kprobe, programName: "trace_ping_v6_sendmsg"},
+		DefaultTcIngress:           &tcProbe{programName: "tc_ingress", tcAttachPoint: bpf.BPFTcIngress},
```
I think these two can be combined into one "probe".
Just use the same fields as we used to have for probes and probeType will be either tcIngress or tcEgress for tc programs.

yanivagman replied:
I mean to combine traceProbe and tcProbe to probe, as it was before. Meaning, to adjust the tc probes to fit the existing probe struct.
No need for direction field. Can simply be probeType (the field name we currently have for probe), with either tcIngress or tcEgress for the tc programs.

rafaeldtinoco replied:
https://github.com/aquasecurity/tracee/pull/1820#issuecomment-1167207980

yanivagman reviewed (commented):
See my comment about the fields usage. I think traceProbe and tcProbe should be combined into one "probe" struct as we used to have, and then the interface can be removed. WDYT?

yanivagman commented:
> > See my comment about the fields usage. I think traceProbe and tcProbe should be combined into one "probe" struct as we used to have, and then the interface can be removed. WDYT?
> 
> Why remove the interface? The code splitting is completely natural here with same function signature and different internal logic. The interface should be made usable (simple line changes in the interface as i've shown), not discarded. In fact the traceProbes should probably be split as well in the future to enable particular capabilities for them (for example the uprobe trigger).

As a future step, I agree, but for now, we already have kprobe, tracepoint, and syscall with different logic to attach the probes, so the natural thing to do as a first step is to simply add the tc probes logic to the existing switch case. Like you say, if we create an interface, then the logical thing to do is to make kprobe, tracepoint and syscall part of this interface as well, but this can be done in a future PR together with tc, or otherwise I think it is confusing

rafaeldtinoco commented:
> > > See my comment about the fields usage. I think traceProbe and tcProbe should be combined into one "probe" struct as we used to have, and then the interface can be removed. WDYT?
> > 
> > 
> > Why remove the interface? The code splitting is completely natural here with same function signature and different internal logic. The interface should be made usable (simple line changes in the interface as i've shown), not discarded. In fact the traceProbes should probably be split as well in the future to enable particular capabilities for them (for example the uprobe trigger).
> 
> As a future step, I agree, but for now, we already have kprobe, tracepoint, and syscall with different logic to attach the probes, so the natural thing to do as a first step is to simply add the tc probes logic to the existing switch case. Like you say, if we create an interface, then the logical thing to do is to make kprobe, tracepoint and syscall part of this interface as well, but this can be done in a future PR together with tc, or otherwise I think it is confusing

I like having them split, like first suggested by Nadav, because tcProbes and tracing Probes are different in nature, but both are "Probes" (i dont like this name by the way, it is misleading since "tc" programs have attach points and hooks). 

The tcProbes need to be attached to interface, need to be detached and attached on-demand, etc (which is something I've considered for my current code I believe).

When coding, tracing probes are very similar, and different than tcprobes, thus the classification in 2 different "Probe" implementations. If we think ahead, we will have the following "Probe" types:

- traceProbes (TYPE_KPROBE, TYPE_TRACEPOINT,, RAW_TRACEPOINT)
- tcProbes (QDISC CLASSIFIERS)
- socketProbes (SOCKET_FILTER, SOCK_OPS, SK_SKB, SK_MSG)
- xdpProbes (express data path)
- tunnelProbes (LWT_IN, LWT_OUT, LWT_XMIT)
- cgroupProbes (CGROUP_SKB, CGROUP_SOCK, CGROUP_DEVICE, CGROUP_SOCKADDR, CGROUP_SYSCTL, CGROUP_SOCKOPT)

and more...

I don't want to go back now to go forward in a near future. I believe addressing concerns from Nadav related to interface implementation are the best next step

If you allow me, I think we should address Nadavs concerns on interface, merge this, allow me to do some more coding regarding network improvements, see if this has a fit or need improvements, etc.

Many of our concerns on classification (for what the interface should look like) will take shape after other probe types appear and we have to implement specific (to their types) functionalities.

rafaeldtinoco commented:
@NDStrahilevitz and I just talked about this one:

- I'm going to fix Probe interface implementation (so it is actually implemented by traceProbe and tcProbe)

- I'll add a fixed "Handle" argument in the methods so it is not part of the variadic "args" approach.

- I'll keep the variadic approach until we have automatic attachment/detachment from network interfaces implemented: this will give us a better idea if we should implement specific arguments or keep the interface generic for future probe types (xdp, cgroups, etc). Then we can make it less generic (variadic) with real usage.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,174 +2,486 @@ package probes
 	}
 
-	return p.probes[probe].attach(p.module)
+	switch b := p.probes[handle].(type) {
```
So since we use (args ...interface{}) Do we actually need this switch? Internally the functions should interpret if and what args are relevant no? I can see leaving it as a safety measure though so up to you.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,174 +2,486 @@ package probes
+	// disable autoload for network related eBPF programs in network is disabled
+	if !netEnabled {
+		for _, in := range allProbes {
+			switch i := in.(type) {
```
Did you remove the part of disabling the network traceProbes on purpose? I remember it being here before.

rafaeldtinoco replied:
I did and there was a comment from Yaniv to have only tc disabled by default in the beginning.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,174 +2,486 @@ package probes
+func (p *probes) attachOrDetach(handle Handle, attach bool, args ...interface{}) error {
+	var netIface *net.Interface
+
+	for _, arg := range args {
```
So if we go by just sending the args forward regardless of concrete probe type, there's no reason to do these conversions, just send the args forward and interpretation will occur internally in the probe.

NDStrahilevitz reviewed (approved):
Left some final style questions anyway, but overall LGTM!

rafaeldtinoco commented:
> Left some final style questions anyway, but overall LGTM!

Yep, I have addressed yours and Yaniv's last comments, I think its much better and cleaner now. A final +1 before merging if nothing else. Thanks for all the reviews.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,175 +2,470 @@ package probes
+//
+
+// enableDisableAutoload enables or disables an eBPF program autoload setting
+func enableDisableAutoload(args ...interface{}) error {
```
Last minute catch, but from reading this function it doesn't currently need the variadic args, maybe we can go back to static types here? Or will there be future probes/hooks where it will be required?

rafaeldtinoco replied:
This one could have static types, I think it will be better!

NDStrahilevitz reviewed (commented):
So I caught one thing last minute, will +1 after, but i'm fine with going ahead with this anyway.

rafaeldtinoco commented:
@yanivagman could I get from you a +1 to merge ? Thanks!

yanivagman reviewed (approved):
LGTM after the changes. Thanks
